homework/hw2-goals-approaches/ez/hw2-ez.md

```diff
@@ -0,0 +1,390 @@
+- load the trained bertopic model into your notebook and display
+    1. the `topic_visualization` interactive plot [see docs](https://maartengr.github.io/BERTopic/api/plotting/topics.html)
+    2. Use the plot to come up with working "names" for each major topic, adjusting the _number_ of topics as necessary to make things more useful. 
+    3. Once you have names, create a _Dynamic Topic Model_ by following [their documentation](https://maartengr.github.io/BERTopic/getting_started/topicsovertime/topicsovertime.html). Use the `release_date` column as timestamps. 
+    4. Describe what you see, and any possible issues with the topic models BERTopic has created. **This is the hardest part... interpreting!**
+
+```{code-cell} ipython3
+from bertopic import BERTopic
+```
+
+```{code-cell} ipython3
+# Load the BERTopic model
+ft_model = BERTopic.load("ft_model")
+```
+
+```{code-cell} ipython3
+# Visualize topics
+ft_model.visualize_topics(top_n_topics = 9)
+```
+
+![dt_model.png](attachment:dt_model.png)
+
+```{code-cell} ipython3
+# Get topic information
+ft_model.get_topic_info()[2:10].set_index("Topic")
+```
+
+#### Renaming Topics
+Topic 1: kami <br>
+Topic 2: goblin <br>
+Topic 3: wildspeaker <br>
+Topic 4: dragon <br>
+Topic 5: darkness <br>
+Topic 6: gerrard <br>
+Topic 7: mage <br>
+Topic 8: temper <br>
+
+```{code-cell} ipython3
+# Prepare data
+df_notna = df.dropna(subset=["flavor_text", "release_date"])
+docs = df_notna.flavor_text.to_list()
+timestamps = df_notna.release_date.to_list()
+```
+
+```{code-cell} ipython3
+# Train a BERTopic model
+topic_model = BERTopic(verbose=True)
+topics, probs = topic_model.fit_transform(docs)
+
+topics_over_time = topic_model.topics_over_time(docs, topics, timestamps)
+```
+
+```{code-cell} ipython3
+# Visualize topics over time
+topic_model.visualize_topics_over_time(topics_over_time, top_n_topics=9)
+```
+
+![topics_over_time.png](attachment:topics_over_time.png)
+
++++
+
+Kami is the Japanese word for "great spirits." From the figure, we can see that the kami topic peaks around 2004, suggesting that there was probably an expansion set about kami released in 2004. After some research, I found out that *Champions of Kamigawa* was released in October 2004 as the first set in the Kamigawa block and it introduced many rare creatures with kami-like magics. I actually started playing mtg recently, and bought the *Kamigawa: Neon Dynasty* expansion set (released in February 2022) over the weekend. According to the storyline, this set represents the current era on Kamigawa, which is more than 1200 years after conclusion of the original Kamigawa block. If we excluded the 2004 kami outlier, we would probably see another smaller spike in 2022.
+
+```{code-cell} ipython3
+# Visualize topics over time without the 'kami' outlier
+topic_model.visualize_topics_over_time(topics_over_time, topics=[0,1,3,4,5,6,7])
```

Comment:
This may be my side only, but locally I'm seeing `kami` as the 0th topic, so this command doesn't remove it.